src/SignalR/clients/csharp/Http.Connections.Client/src/Internal/WebSocketsTransport.cs

```diff
@@ -493,7 +502,7 @@ private async Task StartReceiving(WebSocket socket)
         finally
         {
             // We're done writing
-            if (_gracefulClose)
+            if (_gracefulClose || !_useStatefulReconnect)
```

Comment:
Don't we need to distinguish between if `_gracefulClose` was set to true because we received a WebSocket close message when we might want to reconnect (When does this happen really? Server shutdown? Do we really want to reconnect even then?), an when `StopAsync()` was called by client code and we absolutely don't want to reconnect?

When `HubConnection.StopAsync()` is called, who calls `_application.Output.Complete()` on the transport Pipe?

Reply:
> Don't we need to distinguish between if _gracefulClose was set to true because we received a WebSocket close message when we might want to reconnect (When does this happen really? Server shutdown? Do we really want to reconnect even then?), an when StopAsync() was called by client code and we absolutely don't want to reconnect?

Why do we want to distinguish between those? Both result in closure and no attempt at a reconnect.

> When HubConnection.StopAsync() is called, who calls _application.Output.Complete() on the transport Pipe?

Normally the read loop since `StopAsync` sets `_gracefulClose` to true. But I guess there could be a race where the read loop closes ungracefully and `StopAsync` is called at the same time, so the read loop wouldn't complete the pipe, and then the `ProcessSocketAsync` would exit without completing the pipe too.

Reply:
Added some code to always complete pipes at the end of `ProcessSocketAsync` since it should be a noop to call Complete multiple times.